core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java

```diff
@@ -0,0 +1,928 @@
+   */
+  private class ApplyRules implements Task {
+    private final RelNode mExpr;
+    private final RelSubset group;
+    private final boolean exploring;
+
+    ApplyRules(RelNode mExpr, RelSubset group, boolean exploring) {
+      this.mExpr = mExpr;
+      this.group = group;
+      this.exploring = exploring;
+    }
+
+    @Override public void perform() {
+      Pair<RelNode, Predicate<VolcanoRuleMatch>> category =
+          exploring ? Pair.of(mExpr, planner::isTransformationRule)
+              : Pair.of(mExpr, m -> true);
+      VolcanoRuleMatch match = ruleQueue.popMatch(category);
+      while (match != null) {
+        tasks.push(new ApplyRule(match, group, exploring));
+        match = ruleQueue.popMatch(category);
+      }
+    }
+
+    @Override public void describe(TaskDescriptor desc) {
+      desc.item("mExpr", mExpr).item("exploring", exploring);
+    }
+  }
+
+  /**
+   * APPLY_RULE
+   */
+  private class ApplyRule implements GeneratorTask {
+    private final VolcanoRuleMatch match;
+    private final RelSubset group;
+    private final boolean exploring;
+
+    ApplyRule(VolcanoRuleMatch match, RelSubset group, boolean exploring) {
+      this.match = match;
+      this.group = group;
+      this.exploring = exploring;
+    }
+
+    @Override public void describe(TaskDescriptor desc) {
+      desc.item("match", match).item("exploring", exploring);
+    }
+
+    @Override public void perform() {
+      if (!exploring && !planner.isLogical(match.rel(0)) && !checkLowerBound()) {
+        // for implementation and enforcing rules,
+        // we can skip rule match by checking the lower bound
+        // however, we cannot decide LB for logical nodes,
+        // so we only check enforcing rules now
+        if (!group.getTraitSet().satisfies(match.rels[0].getTraitSet())) {
+          // the match cannot be discarded directly
+          // because of another subset's upper bound
+          ruleQueue.addMatch(match);
+        }
+        return;
+      }
+
+      applyGenerator(this, match::onMatch);
+    }
+
+    private boolean checkLowerBound() {
+      RelOptCost upperBound = group.upperBound;
+      if (upperBound.isInfinite()) {
+        return true;
+      }
+      RelOptCost lb = planner.getLowerBound(match);
+      if (upperBound.isLe(lb)) {
+        if (LOGGER.isDebugEnabled()) {
+          LOGGER.debug(
+              "Skip because of lower bound. LB = {}, UP = {}",
+              lb, upperBound);
+        }
+        return false;
+      }
+      return true;
+    }
+
+    @Override public RelSubset group() {
+      return group;
+    }
+
+    @Override public boolean exploring() {
+      return exploring;
+    }
+  }
+
+  private Task getOptimizeInputTask(RelNode rel, RelSubset group) {
+    if (!rel.getTraitSet().satisfies(group.getTraitSet())) {
+      RelNode passThroughRel = convert(rel, group);
+      if (passThroughRel == null) {
+        if (LOGGER.isDebugEnabled()) {
```

Comment:
The check `if (LOGGER.isDebugEnabled()) {` is redundant.

Reply:
resolved